starknet_py/cairo/serialization/function_serialization_adapter_test.py

```diff
@@ -0,0 +1,78 @@
+from collections import OrderedDict
+
+import pytest
+
+from starknet_py.cairo.serialization.data_serializers.felt_serializer import (
+    FeltSerializer,
+)
+from starknet_py.cairo.serialization.data_serializers.payload_serializer import (
+    PayloadSerializer,
+)
+from starknet_py.cairo.serialization.errors import (
+    InvalidTypeException,
+    InvalidValueException,
+)
+from starknet_py.cairo.serialization.function_serialization_adapter import (
+    FunctionSerializationAdapter,
+)
+from starknet_py.utils.tuple_dataclass import TupleDataclass
+
+payload_serializer = PayloadSerializer(
+    OrderedDict(
+        x=FeltSerializer(),
+        y=FeltSerializer(),
+        z=FeltSerializer(),
+    )
+)
+serializer = FunctionSerializationAdapter(
+    payload_serializer,
+    payload_serializer,
+)
+
+
+def test_serialize():
+    serialized = [1, 2, 3]
+    # pylint: disable=invalid-name
+    x, y, z = serialized
+
+    assert serializer.serialize(x, y, z) == serialized
+    assert serializer.serialize(x=x, y=y, z=z) == serialized
+    assert serializer.serialize(x, y=y, z=z) == serialized
+
+    with pytest.raises(
+        InvalidTypeException, match="Provided 4 positional arguments, 3 possible."
+    ):
+        serializer.serialize(1, 2, 3, 4)
+
+    with pytest.raises(
+        InvalidTypeException,
+        match="Both positional and named argument provided for 'x'.",
+    ):
+        serializer.serialize(1, 2, x=1, y=3)
+
+    with pytest.raises(
+        InvalidTypeException,
+        match="Unnecessary named arguments provided: 'unknown_key'.",
+    ):
+        serializer.serialize(1, 2, unknown_key=3)
+
+    with pytest.raises(InvalidTypeException, match="Missing arguments: 'z'."):
+        serializer.serialize(1, 2)
+
+
+def test_deserialize():
+    assert serializer.deserialize([1, 2, 3]) == TupleDataclass.from_dict(
+        {"x": 1, "y": 2, "z": 3}
+    )
+
+    with pytest.raises(
+        InvalidValueException,
+        match="Error: 2 values out of total 5 values were not used during deserialization.",
```

Comment:
Would it be possible for this error to indicate what values where not used? Like it's done for most other errors?

Reply:
The problem is that there might be a lot of unused values and printing them won't say much if you received them from the API. If you know how many values you've read you can locate the problem in the input.

Reply:
What if we printed values if there were below `n` of them, and truncated otherwise? Like `Values not used [1, 2, 3]` but `Values not used [1, 2, 3, ...] or something along `Values not used [1, 2, 3, ... '12 more follow']`

Reply:
I think the case where someone forgets to (perhaps due to copy paste error) include one or a couple of last values in the abi would be most common.

Reply:
Added truncated info